sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -5425,39 +5424,59 @@ pi_result cuda_piextUSMEnqueueMemcpy2D(pi_queue queue, pi_bool blocking,
       (*event)->start();
     }
 
-    // Determine the direction of Copy using cuPointerGetAttributes
+    // Determine the direction of copy using cuPointerGetAttribute
     // for both the src_ptr and dst_ptr
-    // TODO: Doesn't yet support CU_MEMORYTYPE_UNIFIED
-    CUpointer_attribute attributes = {CU_POINTER_ATTRIBUTE_MEMORY_TYPE};
+    CUDA_MEMCPY2D cpyDesc = {0};
 
     CUmemorytype src_type = static_cast<CUmemorytype>(0);
-    void *src_attribute_values[] = {(void *)(&src_type)};
-    result = PI_CHECK_ERROR(cuPointerGetAttributes(
-        1, &attributes, src_attribute_values, (CUdeviceptr)src_ptr));
-    assert(src_type == CU_MEMORYTYPE_DEVICE || src_type == CU_MEMORYTYPE_HOST);
+    // do not throw if cuPointerGetAttribute returns CUDA_ERROR_INVALID_VALUE
+    // checks with PI_CHECK_ERROR are not suggested
+    CUresult ret = cuPointerGetAttribute(
+        &src_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE, (CUdeviceptr)src_ptr);
+    assert(
+        (src_type != CU_MEMORYTYPE_ARRAY && src_type != CU_MEMORYTYPE_UNIFIED) &&
+        "ARRAY, UNIFIED types are not supported!");
+
+    // pointer not known to the CUDA subsystem (possibly a system allocated ptr)
+    if (ret == CUDA_ERROR_INVALID_VALUE) {
+      cpyDesc.srcMemoryType = CU_MEMORYTYPE_HOST;
+      cpyDesc.srcDevice = 0;
+      cpyDesc.srcHost = src_ptr;
+      // todo: resets the above "non-sticky" error
+    } else if (ret == CUDA_SUCCESS) {
+      cpyDesc.srcMemoryType = src_type;
+      cpyDesc.srcDevice = (src_type == CU_MEMORYTYPE_DEVICE)
+                              ? reinterpret_cast<CUdeviceptr>(src_ptr)
+                              : 0;
+      cpyDesc.srcHost = (src_type == CU_MEMORYTYPE_HOST) ? src_ptr : nullptr;
+    }
```

Comment:
Since this behavior is becoming more complex, would it make sense to split it into a helper function so we can avoid repeating the code? I can see two ways of doing it:

1. Implement it as a function returning a triple, i.e. something like:
```c++
template<typename PtrT>
static std::tuple<CUmemorytype, CUdeviceptr, PtrT> getUSMHostOrDevicePtr(PtrT usm_ptr) { ... }
...
std::tie(cpyDesc.srcMemoryType, cpyDesc.srcDevice, cpyDesc.srcHost) = getUSMHostOrDevicePtr(src_ptr);
...
std::tie(cpyDesc.dstMemoryType, cpyDesc.dstDevice, cpyDesc.dstHost) = getUSMHostOrDevicePtr(dst_ptr);
```
2. Do the same, but make it `void` and pass the members to set as pointers/references:
```c++
template<typename PtrT>
static void getUSMHostOrDevicePtr(PtrT usm_ptr, CUmemorytype *out_mem_type, CUdeviceptr *out_dev_ptr, PtrT *out_host_ptr) { ... }
...
getUSMHostOrDevicePtr(src_ptr, &cpyDesc.srcMemoryType, &cpyDesc.srcDevice, &cpyDesc.srcHost);
...
getUSMHostOrDevicePtr(dst_ptr, &cpyDesc.dstMemoryType, &cpyDesc.dstDevice, &cpyDesc.dstHost);
```
Note I use pointers above because I like to think it makes the intention clearer here, but I do not have a strong preference for one or the other.

Reply:
Sure, I agree.